### DISCUSSION TITLE
OnContactRemoved: last remaining contact?

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#447] [Q&A]
Hi, is there a good way of finding out whether two bodies are still in contact? I'd like to send a callback to the game when a body has fully left a trigger volume, but if I just do that in OnContactRemoved that happens prematurely because multiple contacts get added and removed over time.
I can of course track this on my end, but was just wondering if there is already a nice method I can use.

**Comment by jrouwe** [comment]
There currently is no public API that allows you to detect this, but the information is available inside the physics engine. I need some time to think if this is something that I want to expose / how it could be exposed.

It is basically at:

```
ContactConstraintManager::mCache[mCacheWriteIdx^1].Find(BodyPair(body1, body2), ...) != nullptr
```

but it is not valid at all times.

**Reply by DocAce** [reply]
Could it be a parameter to the function? Eg a bool for whether this was the last contact between the bodies?

**Reply by jrouwe** [reply]
There could, but that would mean that everyone (even if you're not interested in this) would pay with an extra lookup in the contact cache, so that's something I would like to avoid.

**Comment by jrouwe** [comment]
I created a PR #451 that adds a function that allows you to detect if this is the last contact that is being removed, but the function comes with so many caveats that I'm not sure if this is a good API design.

I was wondering: don't you have the same problem when a contact gets added? There you can also get multiple callbacks for the same body and my guess is that you only want to send 1 callback to the game. How do you handle that?

**Reply by DocAce** [reply]
I'm collecting collisions during a simulation step (also when a contact persisted) and calling the game's callback only once with an averaged position.
Oh, and I've added manual tracking of contacts now - sorry, should have updated this thread.
